anki/web/editor.js

```diff
@@ -3,35 +3,88 @@ var rushInOut = (x) => {
     return 2.388 * x - 4.166 * Math.pow(x, 2) + 2.77 * Math.pow(x, 3);
 };
 
-// https://stackoverflow.com/questions/4811822/get-a-ranges-start-and-end-offsets-relative-to-its-parent-container/4812022#4812022
-var cursorOffsetWithin = (element) => {
-    let cursorOffset = 0;
-    let doc = element.ownerDocument || element.document;
-    let win = doc.defaultView || doc.parentWindow;
-    let sel;
-
-    if (typeof win.getSelection != "undefined") {
-        sel = win.getSelection();
-
-        if (sel.rangeCount > 0) {
-            const range = win.getSelection().getRangeAt(0);
-            const preCursortRange = range.cloneRange();
-
-            preCursortRange.selectNodeContents(element);
-            preCursortRange.setEnd(range.endContainer, range.endOffset);
-            cursortOffset = preCursorRange.toString().length;
-        }
-    } else if ((sel = doc.selection) && sel.type != "Control") {
-        const textRange = sel.createRange();
-        const preCursorTextRange = doc.body.createTextRange();
-
-        preCursorTextRange.moveToElementText(element);
-        preCursorTextRange.setEndPoint("EndToEnd", textRange);
-        cursorOffset = preCursorTextRange.text.length;
+/** Python functions moved to JS for async operations **/
+var escape_js_text = (text) => {
+    return text.replace("\\", "\\\\").replace('"', '\\"').replace("'", "\\'");
+}
+
+var replace_or_prefix_old_occlusion_text = (old_html, new_text) => {
+    const occlusion_block_regex = /\[#!occlusions.*?#\]/;
+
+    const new_html = new_text.split(/\r?\n/).join("<br>");
+    replacement = `[#!occlusions ${new_html} #]`;
+
+    /** imitate re.subn **/
+    [subbed, number_of_subs] = ((count = 0) => {
+        const subbed = old_html.replace(occlusion_block_regex, () => {
+            ++count;
+            return replacement;
+        })
+        return [subbed, count];
+    })();
+    
+    if (number_of_subs > 0) {
+        return subbed;
+    } else if (["", "<br>"].includes(old_html)) {
+        return replacement;
+    } else {
+        return `${replacement}<br>${old_html}`;
+    } 
+}
```

Comment:
Why the Python-like formatting?

Reply:
I wanted to make it easier for you to spot the parts that were originally in Python - but with GitLens you'd see the commit message anyway, so there's no valid reason to keep it :D

Reply:
I wanted to make it easier for you to spot the parts that were originally in Python - but with GitLens you'd see the commit message anyway, so there's no valid reason to keep it :D